C++/116_Populating_Next_Right_Pointers_in_Each_Node.cpp

```diff
@@ -0,0 +1,84 @@
+// 116. Populating Next Right Pointers in Each Node
+/**
+ * Given a binary tree
+ * 
+ *     struct TreeLinkNode {
+ *       TreeLinkNode *left;
+ *       TreeLinkNode *right;
+ *       TreeLinkNode *next;
+ *     }
+ *
+ * Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.
+ * 
+ * Initially, all next pointers are set to NULL.
+ * 
+ * Note:
+ * You may only use constant extra space.
+ * You may assume that it is a perfect binary tree (ie, all leaves are at the same level, and every parent has two children).
+ * 
+ * For example,
+ * Given the following perfect binary tree,
+ *
+ *          1
+ *        /  \
+ *       2    3
+ *      / \  / \
+ *     4  5  6  7
+ * 	
+ * After calling your function, the tree should look like:
+ *
+ *          1 -> NULL
+ *        /  \
+ *       2 -> 3 -> NULL
+ *      / \  / \
+ *     4->5->6->7 -> NULL
+ * 
+ * Tags: Tree, Depth-first Search
+ *
+ * Similar Problems: (H) Populating Next Right Pointers in Each Node II, (M) Binary Tree Right Side View
+ * 
+ * Author: Kuang Qin
+ */
+
+#include "stdafx.h"
+
+/**
+ * Definition for binary tree with next pointer.
+ */
+struct TreeLinkNode {
+    int val;
+    TreeLinkNode *left, *right, *next;
+    TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
+};
+
+class Solution {
+public:
+    void connect(TreeLinkNode *root) {
+        while (root != NULL)
+        {
+            // using constant space curr to store current node
+            // this for loop is building a list for the next level, stop if curr has no child
+            for (TreeLinkNode *curr = root; curr->left != NULL; curr = curr->next)
```

Comment:
To be honest, it's confusing. You are moving a pointer to connect a level, while checking whether it has left child in the next level. Seems that the two parts are convoluted. It can be correct, but not that understandable. You should deal with next level in for loop, and each level in while loop, right?

Why not add `curr->left != NULL` to while condition? Then in the for loop, put `curr != NULL` as stop condition. And change the if break to `if(curr->next != NULL) curr->right->next = curr->next->left;`